src/server/ua_server_ns0_gds.c

```diff
@@ -0,0 +1,1111 @@
+                const UA_NodeId *sessionId, void *sessionHandle,
+                const UA_NodeId *methodId, void *methodContext,
+                const UA_NodeId *objectId, void *objectContext,
+                size_t inputSize, const UA_Variant *input,
+                size_t outputSize, UA_Variant *output) {
+
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    /*check for input types*/
+    if(!UA_Variant_hasScalarType(&input[0], &UA_TYPES[UA_TYPES_STRING]) || /*Thumbprint*/
+       !UA_Variant_hasScalarType(&input[1], &UA_TYPES[UA_TYPES_BOOLEAN])) /*IsTrustedCertificate*/
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+
+    UA_String thumbprint = *(UA_String *)input[0].data;
+    UA_Boolean isTrustedCertificate = *(UA_Boolean *)input[1].data;
+
+    /* TODO: If the Certificate is a CA Certificate that has CRLs then all CRLs for that CA are removed as well */
+
+    if(server->transaction)
+        return UA_STATUSCODE_BADTRANSACTIONPENDING;
+
+    UA_CertificateGroup *certGroup = getCertGroup(server, objectId);
+    if(!certGroup)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    /* This Method cannot be called if the containing TrustList Object is open */
+    if(certGroup->fileInfoContext)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    UA_TrustListDataType trustList;
+    memset(&trustList, 0, sizeof(UA_TrustListDataType));
+    trustList.specifiedLists = UA_TRUSTLISTMASKS_ALL;
+
+    UA_ByteString *certificates;
+    size_t certificatesSize = 0;
+    certGroup->getTrustList(certGroup, &trustList);
+
+    if(isTrustedCertificate) {
+        certificates = trustList.trustedCertificates;
+        certificatesSize = trustList.trustedCertificatesSize;
+    } else {
+        certificates = trustList.issuerCertificates;
+        certificatesSize = trustList.issuerCertificatesSize;
+    }
+
+    UA_TrustListDataType list;
+    memset(&list, 0, sizeof(UA_TrustListDataType));
+
+    UA_String thumbpr = UA_STRING_NULL;
+    thumbpr.length = (UA_SHA1_LENGTH * 2);
+    thumbpr.data = (UA_Byte*)malloc(sizeof(UA_Byte)*thumbpr.length);
+    for(size_t i = 0; i < certificatesSize; i++) {
+        UA_CertificateUtils_getThumbprint( &certificates[i], &thumbpr);
+        if(compareThumbprint(&thumbprint, &thumbpr)) {
+            UA_ByteString certificate[1];
+            certificate[0] = certificates[i];
+
+            if(isTrustedCertificate) {
+                list.specifiedLists = UA_TRUSTLISTMASKS_TRUSTEDCERTIFICATES;
+                list.trustedCertificates = certificate;
+                list.trustedCertificatesSize = 1;
+            } else {
+                list.specifiedLists = UA_TRUSTLISTMASKS_ISSUERCERTIFICATES;
+                list.issuerCertificates = certificate;
+                list.issuerCertificatesSize = 1;
+            }
+            break;
+        }
+    }
+
+    if(list.specifiedLists != UA_TRUSTLISTMASKS_NONE) {
+        retval = certGroup->removeFromTrustList(certGroup, &list);
+    } else {
```

Comment:
Even though not explicitly required by the OPC UA specification it seems not be an error to validate the complete trust list again. At least if an issuer certificate is removed here, to ensure, that all remaining certificate chains in the trust list are still intact.